clouddriver-core/src/test/groovy/com/netflix/spinnaker/clouddriver/safety/TrafficGuardSpec.groovy

```diff
@@ -0,0 +1,615 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.safety
+
+import com.netflix.frigga.Names
+import com.netflix.spectator.api.NoopRegistry
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.clouddriver.core.services.Front50Service
+import com.netflix.spinnaker.clouddriver.exceptions.TrafficGuardException
+import com.netflix.spinnaker.clouddriver.model.Cluster
+import com.netflix.spinnaker.clouddriver.model.ClusterProvider
+import com.netflix.spinnaker.clouddriver.model.HealthState
+import com.netflix.spinnaker.clouddriver.model.ServerGroup
+import com.netflix.spinnaker.clouddriver.model.SimpleInstance
+import com.netflix.spinnaker.clouddriver.model.SimpleServerGroup
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import com.netflix.spinnaker.moniker.Moniker
+import retrofit.RetrofitError
+import retrofit.client.Response
+import spock.lang.Ignore
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Unroll
+
+class TrafficGuardSpec extends Specification {
+
+  ClusterProvider clusterProvider = Mock() {
+    getCloudProviderId() >> "aws"
+  }
+
+  Front50Service front50Service = Mock(Front50Service)
+  Registry registry = new NoopRegistry()
+  DynamicConfigService dynamicConfigService = Mock(DynamicConfigService)
+
+  @Shared String location = "us-east-1"
+  @Shared Moniker moniker = new Moniker(app: "app", stack: "foo", cluster: "app-foo", sequence: 1)
+  @Shared String targetName = "app-foo-v001"
+  @Shared String otherName = "app-foo-v000"
+  @Shared Map<String, Object> application = [:]
+
+  @Subject
+  TrafficGuard trafficGuard = new TrafficGuard(
+    Collections.singletonList(clusterProvider),
+    Optional.of(front50Service),
+    registry,
+    dynamicConfigService
+  )
+
+  void setup() {
+    application.clear()
+  }
+
+  def makeServerGroup(String name, int up, int down = 0, Map overrides = [:]) {
+    Set<SimpleInstance> instances = []
+
+    if (up > 0) {
+      instances.addAll((1..up).collect { new SimpleInstance(healthState: HealthState.Up) })
+    }
+    if (down > 0) {
+      instances.addAll((1..down).collect { new SimpleInstance(healthState: HealthState.OutOfService )})
+    }
+
+    ServerGroup serverGroup =  new SimpleServerGroup([
+      region   : 'us-east-1',
+      name     : name,
+      disabled : false,
+      instances: instances,
+      capacity : new ServerGroup.Capacity(min: 0, max: 4, desired: 3)
+    ] + overrides)
+    return serverGroup
+  }
```

Comment:
Tests are all the same except:

1. I `@Ignored` the tests related to `verifyInstanceTermination`
1. I had to refactor `makeServerGroup` because `ServerGroup` has a `Set` for instances which was incompatible with how the tests were originally written. 
1. I also added a `makeCluster` since these things need to typed now instead of maps of maps.
1. I removed one test that was asserting that `namespace` could be used for location. This is unnecessary as the location is explicitly provided via method parameter.